### PR TITLE
Create intra treepath in Engine

### PR DESCRIPTION
**Pull request by ChrisKujawa**
## Description

This PR creates the inter tree-path in the engine and piggybacks it in the ProcessInstanceRecord (in the ACTIVATING EVENT only).

Creating such allows the exporter to reconstruct the tree path, export such to ES, and make use in e.g. Operate.

For more details on why we are building this please see https://github.com/camunda/camunda/issues/25474

### Different Tree paths

There is a difference between inter and intra TreePath. The inter treePath goes over process definitions, including call activities and is used for showing incidents in Operate.

The intra treePath is just inside a process instance, and will/can be used to have a possibility to resolve existing variable scopes and understand the hierarchy in the PI.

We need both versions of treePaths, to support all current features!

### Details: What this PR contains

1. Extend the ProcessInstanceRecord API (and implementation)
    - Allows to piggyback the tree path properties as part of the ACTIVITING event
    - Allows to consume the tree path properties in the exporter - to make it available for the webapps
1. Refactor the tree path builder
    - Remove dependencies to state classes
    - Allows to for easier test set up and less boiler plate code
    - Furthermore, make it easier to test certain cases (e.g. failures)
1. Extend the tree path builder
    - As in the ACTIVATING case no element instance exists yet, we need to pass in the actual flow scope and current RecordValue
3. Add treePath properties to the ProcessInstanceRecord in the ACTIVATING transition
    - During processing the ACTIVATE Command we transition to ACTIVATING (write such event), before this happens we fill the even with necessary details 
    - Doing it in the transition behavior allows us to do it in one place
6. Adjust test cases
    - For tree builder more test cases has been added for flow scope and record value
    - more edge cases have been tested
    - Add tests cases in engine, for processing, that verify tree path is filled in records

### Related issue

relates to https://github.com/camunda/camunda/issues/25474

### CODE REVIEW

File: zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/common/ElementTreePathBuilder.java [hunk 1]
```diff
@@ -7,28 +7,30 @@
...
 import java.util.LinkedList;
 import java.util.List;
 import java.util.Objects;
 
 public class ElementTreePathBuilder {
-  private ElementInstanceState elementInstanceState;
-  private ProcessState processState;
+
+  private ElementInstanceProvider elementInstanceProvider;
+  private CallActivityIndexProvider callActivityIndexProvider;
```
Comment (mustafadagher) [inline zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/common/ElementTreePathBuilder.java:19]: 👍

**Review by mustafadagher — APPROVED** [review]
Great work! 🎉 🚀 
I just have one question below

File: zeebe/protocol-impl/src/main/java/io/camunda/zeebe/protocol/impl/record/value/processinstance/ProcessInstanceRecord.java [hunk 2]
```diff
@@ -59,8 +65,15 @@ public final class ProcessInstanceRecord extends UnifiedRecordValue
   private final LongProperty parentElementInstanceKeyProp =
       new LongProperty("parentElementInstanceKey", -1L);
 
+  private final ArrayProperty<ArrayValue<LongValue>> elementInstancePathProp =
```
[the remark this answers is not in the document]
Comment (ChrisKujawa) [reply zeebe/protocol-impl/src/main/java/io/camunda/zeebe/protocol/impl/record/value/processinstance/ProcessInstanceRecord.java:68]: @mustafadagher no, I wanted to do this in a follow up, after merging https://github.com/camunda/camunda/pull/25459

**Comment by ChrisKujawa** [comment]
@mustafadagher I just found a small bug, already fixed. I added more tests related to call activities, etc. :) 

- [8156bb9](https://github.com/camunda/camunda/pull/25501/commits/8156bb9b5fd02a03628dce13e9df2964bf3ac4e8)
- [df545f8](https://github.com/camunda/camunda/pull/25501/commits/df545f847c7b674839af7ef71bb1105637df88e4)


I need to add the json ignore from the POC. I will do this now.

**Comment by mustafadagher** [comment]
> @mustafadagher I just found a small bug, already fixed. I added more tests related to call activities, etc. :)
> 
> * [8156bb9](https://github.com/camunda/camunda/pull/25501/commits/8156bb9b5fd02a03628dce13e9df2964bf3ac4e8)
> * [df545f8](https://github.com/camunda/camunda/pull/25501/commits/df545f847c7b674839af7ef71bb1105637df88e4)
> 
> I need to add the json ignore from the POC. I will do this now.

yes, good catch with the elementInstanceKey. it was hard to notice.

File: zeebe/protocol-impl/src/main/java/io/camunda/zeebe/protocol/impl/record/value/processinstance/ProcessInstanceRecord.java [hunk 3]
```diff
@@ -169,6 +185,57 @@ public ProcessInstanceRecord setBpmnEventType(final BpmnEventType bpmnEventType)
     return this;
```
Comment (mustafadagher) [inline zeebe/protocol-impl/src/main/java/io/camunda/zeebe/protocol/impl/record/value/processinstance/ProcessInstanceRecord.java:185]: ❌ we should also set the new values in the `wrap()` method

Comment (ChrisKujawa) [reply zeebe/protocol-impl/src/main/java/io/camunda/zeebe/protocol/impl/record/value/processinstance/ProcessInstanceRecord.java:185]: Actually we don't want to pass it around, I would leave this out. I can add a comment that this is on purpose.

**Comment by ChrisKujawa** [comment]
Not sure whether I can solve this with skipping the serialization. The immutable records are always generated with the properties and then the test ExporterSerializationTest fails as the immutable records generate different json than the actual record values..

We might need to include these properties in the json sterilization, that would mean we need to extend the ES schema..

**Comment by ChrisKujawa** [comment]
I tried several things, to work around this issue see https://github.com/camunda/camunda/tree/ck-try-optional, but at the end nothing worked well enough.

I went the way we did for the incident record and added to the schema but disabled by default.

File: zeebe/protocol-impl/src/main/java/io/camunda/zeebe/protocol/impl/record/value/processinstance/ProcessInstanceRecord.java [hunk 4]
```diff
@@ -186,7 +186,6 @@ public ProcessInstanceRecord setBpmnEventType(final BpmnEventType bpmnEventType)
   }
 
   @Override
-  @JsonIgnore
```
[the remark this answers is not in the document]
Comment (ChrisKujawa) [reply zeebe/protocol-impl/src/main/java/io/camunda/zeebe/protocol/impl/record/value/processinstance/ProcessInstanceRecord.java]: It never stops.

[the remark this answers is not in the document]
Comment (mustafadagher) [reply zeebe/protocol-impl/src/main/java/io/camunda/zeebe/protocol/impl/record/value/processinstance/ProcessInstanceRecord.java]: I know :D

File: zeebe/protocol-impl/src/main/java/io/camunda/zeebe/protocol/impl/record/value/processinstance/ProcessInstanceRecord.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (mustafadagher) [reply zeebe/protocol-impl/src/main/java/io/camunda/zeebe/protocol/impl/record/value/processinstance/ProcessInstanceRecord.java:68]: @ChrisKujawa have we ever re-planned this?

[the remark this answers is not in the document]
Comment (ChrisKujawa) [reply zeebe/protocol-impl/src/main/java/io/camunda/zeebe/protocol/impl/record/value/processinstance/ProcessInstanceRecord.java:68]: Thanks for the ping. No it was more a performance related follow up, that i havent seen as must have.

[the remark this answers is not in the document]
Comment (ChrisKujawa) [reply zeebe/protocol-impl/src/main/java/io/camunda/zeebe/protocol/impl/record/value/processinstance/ProcessInstanceRecord.java:68]: I have to check whether i have an issue for it already

**Pull request merged**